cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/join/JoinCondition.java

```diff
@@ -0,0 +1,223 @@
+ *
+ * Currently joins can only be performed on equality on a set of fields from each stage.
+ */
+public class JoinCondition {
+  private final Op op;
+
+  private JoinCondition(Op op) {
+    this.op = op;
+  }
+
+  public Op getOp() {
+    return op;
+  }
+
+  /**
+   * Validate that is condition is valid to use when joining the given stages.
+   *
+   * @param joinStages the stages that will be joined on this condition
+   * @throws InvalidJoinConditionException if the condition is invalid
+   */
+  public void validate(List<JoinStage> joinStages) {
+    // no-op
+  }
+
+  /**
+   * Condition operation.
+   */
+  public enum Op {
+    EQUAL_TO
+  }
+
+  public static OnKeys.Builder onKeys() {
+    return new OnKeys.Builder();
+  }
+
+  /**
+   * Join on multiple keys from each stage.
+   */
+  public static class OnKeys extends JoinCondition {
+    private final Set<JoinKey> keys;
+    private final boolean dropNullKeys;
+
+    private OnKeys(Set<JoinKey> keys, boolean dropNullKeys) {
+      super(Op.EQUAL_TO);
+      this.keys = Collections.unmodifiableSet(new HashSet<>(keys));
+      this.dropNullKeys = dropNullKeys;
+    }
+
+    public Set<JoinKey> getKeys() {
+      return keys;
+    }
+
+    public boolean shouldDropNullKeys() {
+      return dropNullKeys;
+    }
+
+    @Override
+    public void validate(List<JoinStage> joinStages) {
+      super.validate(joinStages);
+
+      Map<String, JoinStage> stageMap = joinStages.stream()
+        .collect(Collectors.toMap(JoinStage::getStageName, s -> s));
+
+      for (JoinKey joinKey : keys) {
+        String joinStageName = joinKey.getStageName();
+        JoinStage joinStage = stageMap.get(joinStageName);
+        // check that the stage for each key is in the list of stages
+        if (joinStage == null) {
+          throw new InvalidJoinConditionException(String.format(
+            "Join key for stage '%s' is invalid. Stage '%s' is not a join input.",
+            joinStageName, joinStageName));
+        }
+        // this happens if the schema for that stage is unknown.
+        // for example, because of macros could not be evaluated yet.
+        if (joinStage.getSchema() == null) {
+          continue;
+        }
+        Set<String> fields = joinStage.getSchema().getFields().stream()
+          .map(Schema.Field::getName)
+          .collect(Collectors.toSet());
+
+        // check that the key fields for each key is in fields for that stage
+        // for example, when joining on A.id = B.uid, check that 'id' is in the fields for stage A and 'uid' for stage B
+        Set<String> keysCopy = new HashSet<>(joinKey.getFields());
+        keysCopy.removeAll(fields);
+        if (keysCopy.size() == 1) {
+          throw new InvalidJoinConditionException(String.format(
+            "Join key for stage '%s' is invalid. Field '%s' does not exist in the stage.",
+            joinStageName, keysCopy.iterator().next()));
+        } else if (keysCopy.size() > 1) {
```

Comment:
No need to have `else` after a throw inside the `if` block.

Reply:
this is an else-if, since it's also possible the set is empty, in which case it's valid.

Reply:
I mean it's cleaner to write:

```
if (keyCopy.size() == 1) {
  throw
}
if (keyCopy.size() > 1) {
  throw
}
```